src/Compilers/CSharp/Portable/Parser/LanguageParser.cs

```diff
@@ -6266,13 +6335,117 @@ private TypeSyntax ParseUnderlyingType(bool parentIsParameter, NameOptions optio
             {
                 return this.ParseTupleType();
             }
+            else if (IsFunctionPointerStart())
+            {
+                return ParseFunctionPointerTypeSyntax();
+            }
             else
             {
                 var name = this.CreateMissingIdentifierName();
                 return this.AddError(name, ErrorCode.ERR_TypeExpected);
             }
         }
 
+#nullable enable
+        private FunctionPointerTypeSyntax ParseFunctionPointerTypeSyntax()
+        {
+            Debug.Assert(IsFunctionPointerStart());
+            var @delegate = EatToken(SyntaxKind.DelegateKeyword);
+            var asterisk = EatToken(SyntaxKind.AsteriskToken);
+            // Invalid calling conventions will be reported at type binding time
+            var callingConvention = IsPossibleCallingConvention() ? EatTokenAsKind(SyntaxKind.IdentifierToken, reportError: false) : null;
+
+            if (!IsPossibleFunctionPointerParameterListStart(CurrentToken))
+            {
+                var lessThanTokenError = SyntaxFactory.MissingToken(SyntaxKind.LessThanToken);
+                lessThanTokenError = WithAdditionalDiagnostics(lessThanTokenError, GetExpectedTokenError(SyntaxKind.LessThanToken, SyntaxKind.None));
+                var typesError = _pool.AllocateSeparated<FunctionPointerParameterOrReturnTypeSyntax>();
+                var identifierError = CreateMissingIdentifierName();
+                identifierError = AddError(identifierError, ErrorCode.ERR_TypeExpected);
+                var typeError = SyntaxFactory.FunctionPointerParameterOrReturnType(default, identifierError);
+                typesError.Add(typeError);
+                var greaterThanTokenError = SyntaxFactory.MissingToken(SyntaxKind.GreaterThanToken);
+                greaterThanTokenError = WithAdditionalDiagnostics(greaterThanTokenError, GetExpectedTokenError(SyntaxKind.GreaterThanToken, SyntaxKind.None));
+
+                var funcPtr = SyntaxFactory.FunctionPointerType(@delegate, asterisk, callingConvention, lessThanTokenError, typesError, greaterThanTokenError);
+                _pool.Free(typesError);
+                return funcPtr;
+            }
+
+            var lessThanToken = EatToken(SyntaxKind.LessThanToken);
+            var saveTerm = _termState;
+            _termState |= TerminatorState.IsEndOfFunctionPointerParameterList;
+            var types = _pool.AllocateSeparated<FunctionPointerParameterOrReturnTypeSyntax>();
+
+            try
+            {
+                while (true)
+                {
+                    var modifiers = _pool.Allocate<SyntaxToken>();
+                    try
+                    {
+                        // Modifiers will be checked for validity in later steps
+                        while (IsParameterModifier(CurrentToken.Kind) || CurrentToken.Kind == SyntaxKind.ReadOnlyKeyword)
```

Comment:
why not just ParseModifiers?  then you can flexibly handle user mistakes and report back a standard error for it using our modifier-checking code?

Reply:
Because I didn't see that method :)

---
In reply to: [348865046](https://github.com/dotnet/roslyn/pull/39879#discussion_r348865046) [](ancestors = 348865046)

Reply:
Note: after investigation, none of the existing ParseXModifiers code paths work quite correctly here, and it's not worth the effort to share code with ParseParameterModifiers here because the majority of that function is just checking to see if language versions are set.a

---
In reply to: [349262113](https://github.com/dotnet/roslyn/pull/39879#discussion_r349262113) [](ancestors = 349262113,348865046)